src/librustc/hir/lowering.rs

```diff
@@ -5214,3 +5235,12 @@ fn body_ids(bodies: &BTreeMap<hir::BodyId, hir::Body>) -> Vec<hir::BodyId> {
     body_ids.sort_by_key(|b| bodies[b].value.span);
     body_ids
 }
+
+fn returns_impl_trait(decl: &FnDecl) -> bool {
+    if let FunctionRetTy::Ty(ref ty) = decl.output {
+        if let TyKind::ImplTrait(..) = ty.node {
```

Comment:
I think we should also support existential types here. Existential types should be detectable by checking which definition a `TyKind::Path` refers to. Might require a lowered function signature